connectors/sync_job_runner.py

```diff
@@ -0,0 +1,167 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+import asyncio
+import time
+
+from connectors.byoc import JobStatus
+from connectors.es import Mappings
+from connectors.filtering.validation import (
+    FilteringValidationState,
+    InvalidFilteringError,
+)
+from connectors.logger import logger
+
+
+class SyncJobRunningError(Exception):
+    pass
+
+
+class SyncJobRunner:
+    """The class to run a sync job."""
+
+    def __init__(
+        self,
+        source_klass,
+        sync_job,
+        connector,
+        elastic_server,
+        bulk_options,
+    ):
+        self.source_klass = source_klass
+        self.sync_job = sync_job
+        self.job_id = self.sync_job.id
+        self.connector = connector
+        self.connector_id = self.connector.id
+        self.elastic_server = elastic_server
+        self.bulk_options = bulk_options
+        self._start_time = None
+        self.running = False
+
+    async def execute(self):
+        if self.running:
+            raise SyncJobRunningError(f"Sync job #{self.job_id} is already running.")
+
+        self.running = True
+        if not await self._claim_job():
+            logger.error(f"Unable to claim job #{self.job_id}")
+            return
+
+        try:
+            self._start_time = time.time()
+            sync_status = None
+            sync_error = None
+
+            data_provider = self.source_klass(self.sync_job.configuration)
+            if not await data_provider.changed():
+                logger.debug(
+                    f"No change in {self.sync_job.service_type} data provider, skipping..."
+                )
+                return
+
+            logger.debug(f"Syncing '{self.sync_job.service_type}'")
+            logger.debug(f"Pinging the {self.source_klass} backend")
+            await data_provider.ping()
+            await asyncio.sleep(0)
```

Comment:
you can remove this one because you have another await right after.

```
await asyncio.sleep(0)
```

is useful in a block of sync calls that loops, to return execution to the event loop, and that happens on any await call